### DISCUSSION TITLE
Does every git user need git-lfs?

### DISCUSSION THREAD

**Discussion by mdelp-es** [git-lfs/git-lfs#4644] [Q&A]
If I have files in the github repo tracked with git lfs, then does every user of the repository need git lfs installed?

(I assume the answer is yes, but just checking)

**Comment by chrisd8088** [comment]
I think the answer depends a lot on how your repository is structured and what the workflow of your collaborators is expected to be.

Suppose, for instance, that you only use Git LFS to manage image files, and all you put all the image files under an `/images/` folder.  You might then use `git lfs track '/images/*'` to establish a rule in `.gitattributes` that tells Git LFS's filters to only convert files in that folder (and any subfolders) to LFS pointers, but to convert _all_ files in that folder, on the assumption they are all image files and you want to use LFS to track all such files.

Suppose you also have a bunch of source code you manage in a folder named `/src/`.  If someone is only working on the source code and never needs to look at or alter the images files, they might successfully skip using Git LFS and still be able to commit to your repository.  If, however, they ventured into the `/images/` folder, they'd see files with contents like:
```
version https://git-lfs.github.com/spec/v1
oid sha256:02ad5a3e31dddc993efb45d1f4207540d125f14c9e22e79d325c8fc3554e3520
size 7652
```

That's a Git LFS pointer file, which replaces the image file in the Git repository.  Without Git LFS installed, it will just appear as a "bare" pointer file.  With Git LFS installed, the filter hook programs would run when you checked out the branch, pointer files would be detected and the relevant actual file downloaded from the Git LFS server and put in place instead of the pointer file in your working tree.

For more advanced users, you can also install Git LFS, but then tell it not to download the actual files but just to leave the pointer files in place with the `GIT_LFS_SKIP_SMUDGE=1` environment variable, or in a more fine-grained manner with various specialized configuration options like `lfs.fetchexclude` or options to certain Git LFS commands like `git lfs fetch --exclude`.  It's definitely worth reading the `git-lfs-config(5)` [man page](https://github.com/git-lfs/git-lfs/blob/main/docs/man/git-lfs-config.5.ronn#fetch-settings) and the other man pages, and experimenting yourself with various settings.

**Reply by amartani** [reply]
@chrisd8088 from my testing, I believe that every user must at least have the git-lfs binary installed, otherwise, their checkout will fail with:

`git-lfs filter-process: git-lfs: command not found`

Is there a way to make the checkout process not fail without the binary being installed?

**Comment by chrisd8088** [comment]
Is it possible that error derives from having something like the following in your `$HOME/.gitconfig` or system-level Git configuration files?
```
[filter "lfs"]
       process = git-lfs filter-process
       required = true
       clean = git-lfs clean -- %f
       smudge = git-lfs smudge -- %f
```

If I comment those out of my `~/,gitconfig`, and temporarily rename the `git-lfs` binary so it appears as if it's not installed, then I can cleanly clone and checkout a Git repo that contains Git LFS pointer files.  You might try testing on a system which has never had Git LFS installed, for instance.  The thing is, there's no magic in Git itself which would recognize Git LFS objects and report an error.  The repo is just a Git repo where large files happen to be replaced with small files containing Git LFS pointer data.

All the "magic" happens via Git's filters which have to be configured to kick off Git LFS processes during certain operations.  If you don't have those Git configurations, nothing LFS-related happens, even if the `git-lfs` binary is on the system.  In fact, I can leave the `git-lfs` binary installed but comment out my `lfs` block of Git configurations, and a clone will just give me the bare Git repo with the Git LFS pointer files in it, unconverted to regular files.

**Comment by SalmaFarghaly** [comment]
Hi @chrisd8088 ,
I have a question regarding that If the locking mechanism has been applied to a repo and Then some other user has cloned the repo but doesn't have GIT LFS in his system and let's assume he is one of the repo's contributors. Will he be able to edit the lockable files without taking a lock? and Will the lockable files be read/write in his workspace after the clone?

**Comment by chrisd8088** [comment]
I believe the short answers are yes, and yes.  The whole locking mechanism is advisory and is implemented by Git LFS, not Git, so if you just do a regular Git clone then pushing and pulling work as normal, and Git LFS is never consulted.  You have to have the Git LFS client installed in order for its filter programs to be invoked by Git, check for locks, and report their existence.  Otherwise it's just a normal Git repo underneath.  The same applies to Git LFS pointer files -- if you don't have Git LFS installed, you just see the pointer files in the repo.
